DREAM/Optimization/tsgParticleSwarm.cpp

```diff
@@ -0,0 +1,168 @@
+
+    std::vector<double> box_particle_velocities(num_dimensions * box_num_particles);
+    std::generate(box_particle_velocities.begin(), box_particle_velocities.end(), get_random01);
+    std::copy(box_particle_velocities.begin(), box_particle_velocities.end(), std::back_inserter(particle_velocities));
+}
+
+void ParticleSwarm(BatchedObjectiveFunction f, int max_iterations, TasDREAM::DreamDomain inside, ParticleSwarmState &state,
+                   double inertia_weight, double cognitive_coeff, double social_coeff, std::function<double(void)> get_random01) {
+
+    // Initialize.
+    size_t num_dimensions = (size_t) state.getNumDimensions();
+    size_t num_particles = (size_t) state.getNumParticles();
+    std::vector<double> particle_positions = state.getParticlePositions();
+    std::vector<double> particle_velocities = state.getParticleVelocities();
+    std::vector<double> best_particle_positions = state.getBestParticlePositions();
+    if (best_particle_positions.size() == 0) {
+        best_particle_positions = particle_positions;
+        state.setBestParticlePositions(best_particle_positions);
+    }
+    std::vector<double> best_position = state.getBestPosition();
+    std::vector<double> prev_fvals(num_particles), fvals(num_particles), candidate(num_dimensions);
+    double best_fval = std::numeric_limits<double>::max();
+    if (best_position.size() == 0) {
+        f(particle_positions, fvals);
+        for (size_t i=0; i<num_particles; i++) {
+            std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                      candidate.begin());
+            if (inside(candidate) && fvals[i] < best_fval) {
+                best_position = candidate;
+                best_fval = fvals[i];
+            }
+        }
+        state.setBestPosition(best_position);
+    }
+    f(best_particle_positions, prev_fvals);
+
+    // Main optimization loop.
+    while(state.getNumIterations() < max_iterations) {
+        for (size_t i=0; i<num_particles; i++) {
+            for (size_t j=0; j<num_dimensions; j++) {
+                double rp = get_random01();
+                double rg = get_random01();
+                int idx = i * num_dimensions + j;
+                particle_velocities[idx] = inertia_weight * particle_velocities[idx] +
+                                           cognitive_coeff * rp * (best_particle_positions[idx] - particle_positions[idx]) +
+                                           social_coeff * rg * (best_position[j] - particle_positions[idx]);
+            }
+            for (size_t j=0; j<num_dimensions; j++) {
+                particle_positions[i * num_dimensions + j] += particle_velocities[i * num_dimensions + j];
+            }
+        }
+        state.setParticlePositions(particle_positions);
+        state.setParticleVelocities(particle_velocities);
+        // Update the best positions.
+        f(particle_positions, fvals);
```

Comment:
Here, `f()` may be called with points that don't satisfy `inside()`.

We should have two methods here, one that uses an `inside` function and one that is unconstrained. The one that uses `inside()` should make a second check here and call `f()` only for the points that satisfy the constraint.

Reply:
I'm not worried about `fvals`, I'm worried about `particle_positions`. The objective shouldn't have to sanitize the input, if the user says "consider this constraint" then the constraint should be respected. If the objective relies on a model or a surrogate that has limited area of validity, our algorithm should respect that.